### DISCUSSION TITLE
Error handling in plugin if address already in use

### DISCUSSION THREAD

**Discussion by pschaub** [fastify/fastify#4183] [General] [answered]
### Scenario

- During startup our plugin `plugins/prisma.ts` is connecting to the database server.
- During `fastify.listen()` an error happens: `Error: listen EADDRINUSE: address already in use 127.0.0.1:8080`

### Issue
The database connection is already initialized. How can we catch the occured error inside the plugin to disconnect from the database?
There should be a hook to catch initialization errors, e.g. `onListenError`?

### Example

plugins/prisma.ts
```ts
import fp from 'fastify-plugin'
import { FastifyPluginAsync } from 'fastify'
import { PrismaClient } from '@prisma/client'

declare module 'fastify' {
    interface FastifyInstance {
        prisma: PrismaClient
    }
}

const prismaPlugin: FastifyPluginAsync = fp(async (server, options) => {
    const prisma = new PrismaClient();

    await prisma.$connect();

    server.decorate('prisma', prisma);

    server.addHook('onClose', async (server) => {
        await server.prisma.$disconnect();
    })
})

export default prismaPlugin
```

**Comment by mcollina** [comment, marked as answer]
I think you can call `await fastify.close()`. Would you like to give it a try?

**Reply by pschaub** [reply]
@mcollina so your solution would be this?

```ts
await fastify.listen({port: 8080}).catch(async (err) => {
    await fastify.close();
    throw err;
});
```

I tried it and it works. But it feels wrong that I need to write this code.

I am sure a lot of developers just copy the plugin code and never think about this case at all.
The above code is an example on this page: https://www.prisma.io/fastify

What do you think about calling `await fastify.close()` inside fastify itself if an error occurs during `listen`?
Or how about adding a new hook for this situation?

**Reply by mcollina** [reply]
The code in the prisma doc is correct, in the sense that the process will crash with an error.

If you want to handle said error yourself (with a `.catch()`), then you would need to close the connection or do anything. This is hardly newbie behavior - in case of a fatal error the best answer is almost always to crash.
